hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,409 @@
+# %%
+import warnings
+from collections import Counter, defaultdict
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from .._ipcw import KaplanMeierIPCW
+from ..utils import check_y_survival
+
+
+def concordance_index_incidence(
+    y_test,
+    y_pred,
+    time_grid=None,
+    taus=None,
+    y_train=None,
+    X_train=None,
+    X_test=None,
+    event_of_interest=1,
+    ipcw_estimator="km",
+    tied_tol=1e-8,
+):
```

Comment:
For the API, maybe instead of passing `y_train, X_train, ipcw_estimator` we could have instead a single required `censoring_model` argument. Then it's up to the user to pass a fitted estimator $\hat{G}(t)$ (Kaplan-Meier) or $\hat{G}(t|x)$ (Cox regression, the one learned by SurvivalBoost)  or even the oracle $G^*(t|x)$ for synthetic data. The same could be done for `brier_score_survival` and friends.

```suggestion
def concordance_index_incidence(
    y_test,
    y_pred,
    censoring_model,
    time_grid=None,
    taus=None,
    X_test=None,
    event_of_interest=1,
    tied_tol=1e-8,
):
```

Reply:
Yes, I like this idea. Should we also allow the user to pass a string like "kaplan-meier" to keep this function easy to use while providing extensions with a custom estimator? So we would accept this string and $\hat{G}$ estimators

Reply:
To avoid slowing down this PR further, I suggest we only support None and "km" for now. The mechanism to use conditional IPCW is not there yet. So, I'll remove X_test and X_train from the function parameters.

Reply:
Sounds good to me. Allowing any estimator for the censoring model should probably be done in a specific PR.